documentation.lisp

```diff
@@ -147,6 +147,9 @@ same goes for calling MUNMAP with values not directly returned by MMAP,
 calling it with changed values returned by MMAP, or attempting to
 dereference the PTR after a call to MUNMAP.
 
+If an open file descriptor was passed to MMAP, then MUNMAP will not attempt
+to close the file.
```

Comment:
As discussed in chat, please do not implement it like this. Instead note in MUNMAP that you may pass NIL to forego closure of the FD, and always return a valid FD from MMAP.

Reply:
Ahh right, I forgot about that. 🤦‍♂️

That said, I'm not sure how to preserve the fd return value without also closing the file or in some form communicating that the fd ought to be left open.

As you noted, passing in NIL forgoes closure of the fd. This in combination with fd being an opaque value, no other functions using fd, and the caller clearly knowing the fd to be able to pass it in, presents to me that this is the most ergonomic way to do this.

Are you looking for the logic for close-skipping to live in a higher level function, like with-mmap?

Reply:
My argument is this: if I pass in an FD, I would expect to get the FD back out. It makes no sense to get a "fresh FD" even if it is an opaque value. I may also want to close the FD along with unmapping, so specifying that behaviour in MUNMAP seems reasonable.

As for WITH-MMAP, we could add a keyword argument DONT-CLOSE which will pass NIL to MUNMAP in the unwind.

Reply:
I took another look at WITH-MMAP. Since its paramlist ends with a &REST, I don't believe there's backwards-compatible room for a keyword to indicate not to close.

I *could* add a check to nil-punt the fd if PATH is an int, but that would just bring the same surprising-behavior issue here. Another solution could be to add a WITH-MMAP* to either opt not to close, or to make space for keyword options, though, *euugh*, I really don't like the idea of exposing a second macro doing (almost) the same thing. What are your thoughts?

Reply:
There is a backwards compatible way. Just add a `&key dont-close` and `(remf args :dont-close)` in the body before passing it on. The key so far was not used in any backends and as such is completely fine to add.

Reply:
Fixed, thanks :+1: